### PR TITLE
Match the vision model_type get_model_type actually returns in VLM layer counts

### PR DESCRIPTION
**Pull request by danielhanchen**
## Summary

`get_model_layer_counts` branched on the top level `config.model_type`, but `get_model_type` deliberately prefers `vision_config.model_type` (vLLM reports Qwen2.5-VL as `qwen2_5_vl_text`, so the vision name is the reliable one). The spellings never lined up, so several branches were unreachable and the counts fell through to the causal-LM default of 32.

`extract_vision_layers` uses `max(layer_counts.values())` purely as an iteration bound, and `_get_nested_attr` returns `None` past the end of a `ModuleList`. Over-counting is therefore harmless, but under-counting silently skips every weight above index 31 with no error.

## Before and after

Measured with `AutoConfig`, so these are the strings the code actually sees:

| model | `get_model_type` | before | after | real depth |
|---|---|---|---|---|
| Llama-3.2-11B-Vision | `mllama_vision_model` | 32 | 40 | 40 |
| Llama-3.2-90B-Vision | `mllama_vision_model` | 32 | 100 | 100 |
| gemma-3-4b-it | `siglip_vision_model` | 32 | 34 | 34 |
| gemma-3-27b-it | `siglip_vision_model` | 32 | 62 | 62 |
| gemma-4-E2B-it | `gemma4_vision` | 32 | 35 | 35 |
| gemma-4-31B-it | `gemma4_vision` | 32 | 60 | 60 |
| gemma-4-12B-it | `gemma4_unified_vision` | 32 | 48 | 48 |
| Qwen3-VL-8B (transformers 5.17) | `qwen3_vl_vision` | 32 | 36 | 36 |
| Qwen2.5-VL, Idefics3, SmolVLM2, plain causal LMs | unchanged | - | - | - |

The concrete loss is on Mllama, whose `additional_layers` templates are indexed by **text** layer:

```
model.language_model.model.layers.{kk}.cross_attn_attn_gate
model.language_model.model.layers.{kk}.cross_attn_mlp_gate
```

Llama-3.2-11B-Vision has `cross_attention_layers = [3, 8, 13, 18, 23, 28, 33, 38]`. Stopping at 32 drops layers 33 and 38, so 4 of 16 gate tensors were never extracted. On the 90B, 16 of 20 cross-attention layers sit above 31, dropping 32 tensors. Driving `extract_vision_layers` over a synthetic Mllama tree confirms 12/16 gate keys before and 16/16 after, with nothing lost.

Gemma 3 and Gemma 4 were latent rather than broken: their vision towers are 27 layers or fewer, so the accidental 32 happened to cover them. They break the moment a deeper vision tower ships.

## Changes

**`empty_model.py`**
- `get_model_layer_counts`: accept the vision spellings alongside the top level ones for `mllama`, `gemma3`, `gemma4` (both the `gemma4` and `gemma4_unified` families) and `qwen3_vl`
- read `text_config` / `vision_config` through `getattr(config, ..., config)`, so a config carrying a vision `model_type` but no sub-configs returns counts instead of raising `AttributeError`
- hoist the Qwen merged-qkv list into `QWEN_VL_MERGED_QKV_TYPES`, shared by the three places that gate on it, so they cannot drift apart again

**`vllm_utils.py`**
- add `model.text_model.lm_head.weight` to the tied lm_head exclusion, as `TIED_LM_HEAD_KEYS`. This is the Idefics3 / SmolVLM2 layout where lm_head is tied to the text model embeddings
- the exclusion was one-directional: the key was excused from the symmetric-difference check, then the per-key compare looked it up in the other dict unconditionally and raised `KeyError` two lines later. The fallback now runs for every tied lm_head name, against `TIED_EMBED_KEYS`

## transformers 5 renamed the Qwen3-VL vision config

`qwen3_vl` -> `qwen3_vl_vision`, which silently took that branch out. Checked across three versions in isolated venvs:

| | 4.57.6 | 5.0.0 | 5.17.0 |
|---|---|---|---|
| Qwen3-VL `vision_config.model_type` | `qwen3_vl` | `qwen3_vl` | `qwen3_vl_vision` |
| Qwen3-VL bound before | 36 | 36 | 32 |
| Qwen3-VL bound after | 36 | 36 | 36 |

This also affects the merged-qkv decision in `extract_vision_layers`: on 5.17 the branch was missed, so a qkv that HF keeps merged got split into `q_proj`/`k_proj`/`v_proj`. Accepting both spellings restores the 4.57.6 behaviour rather than introducing new behaviour. Mllama and Gemma 3 report the same strings on all three versions, so that part of the fix is version independent.

## Compatibility

The change only reads config attributes and does set and dict arithmetic, so it is independent of torch, TRL, PEFT and vLLM versions, and has no platform-specific code. Verified on transformers 4.57.6 (torch 2.6.0), 5.0.0 and 5.17.0.

Nothing narrows: every widened condition is a superset of what matched before, and the branch bodies are unchanged. `getattr(config, "text_config", config)` only changes outcomes where the old code raised. Qwen2.5-VL, Idefics3, SmolVLM2 and plain causal LMs come out byte-identical.

## Cross-platform staging CI

Run off the org queue on `danielhanchen/unsloth-zoo-staging`, branch `pr-550-xplat-ci`:

| runner | result |
|---|---|
| ubuntu-latest | 99 passed |
| windows-latest | 99 passed |
| macos-15 (Apple Silicon) | not applicable, skipped at collection |

macOS is a skip rather than a pass, and deliberately so: `pyproject.toml` excludes torch on
`darwin/arm64` in favour of MLX, and the code this PR touches is the torch and vLLM weight
conversion path, which does not exist on the MLX side. The staging job confirms that shows up
as a skip rather than a failure. Linux and Windows are where these tests can actually execute,
and both run the full 99.

## Test plan

- [x] `tests/test_vlm_layer_counts.py`, 30 cases: both spellings per family, the Mllama end-to-end extraction asserting the layer 33 and 38 gates are present, plain causal LMs still returning an `int`, missing sub-configs not raising, and an unknown vision model still falling through
- [x] `tests/test_tied_lm_head_state_dict.py`, 6 cases: every tied lm_head name absent from either side, and still reported when both sides carry it with values that disagree
- [x] Both files fail on the merge base and pass here
- [x] Genuine key and value mismatches still raise, so the relaxed exclusion did not weaken the check
- [x] Layer counts checked against `AutoConfig` for 15 real checkpoints across Mllama, Gemma 3, Gemma 4, Qwen2.5-VL, Qwen3-VL, Idefics3, SmolVLM2 and text-only models
- [x] A tied lm_head present only in the converted state dict is now compared against the
      embedding it is tied to. Previously it was excused from the difference check and then
      never compared, so a converted head disagreeing with its embedding validated clean;
      this was already true of `lm_head.weight` and `model.language_model.lm_head.weight`
      before this PR. Covered in both directions for all four tied names
- [x] Existing suite: same results as the merge base. `test_checkpoint_compile_mix.py::test_settlement_retries_until_the_generator_is_actually_collected` fails identically before and after and is unrelated

### CODE REVIEW

File: unsloth_zoo/vllm_utils.py [hunk 1]
```diff
@@ -1190,7 +1190,7 @@ def assert_same_state_dict(old_state_dict, new_state_dict):
     # hf, vllm
 
     difference = new_state_dict.keys() ^ old_state_dict.keys()
-    difference -= set(("model.lm_head.weight","model.language_model.lm_head.weight", "lm_head.weight"))
+    difference -= set(("model.lm_head.weight", "model.language_model.lm_head.weight", "model.text_model.lm_head.weight", "lm_head.weight"))
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/vllm_utils.py]: Done, switched to a set literal. It is now a module level `TIED_LM_HEAD_KEYS` frozenset, since the per-key comparison loop further down needs the same names: the key was excused from the symmetric-difference check and then looked up in the other state dict unconditionally, so it raised `KeyError` a few lines later when it was only present on one side. Sharing one set fixes that direction too and keeps the two places from drifting.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

File: unsloth_zoo/vllm_utils.py [hunk 2]
```diff
@@ -1344,7 +1358,7 @@ def _normalize_state_dict_tensor(value):
         return value.contiguous()
 
     difference = new_state_dict.keys() ^ old_state_dict.keys()
-    difference -= set(("model.lm_head.weight","model.language_model.lm_head.weight", "lm_head.weight"))
+    difference -= TIED_LM_HEAD_KEYS
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/vllm_utils.py]: Confirmed and fixed in 5d50cc24. Reproduced it first: with old={model.text_model.embed_tokens.weight} and new={that embedding, model.text_model.lm_head.weight=different}, assert_same_state_dict returned clean, while the same mismatch in the old-only direction raised. The subtraction empties the symmetric difference and the per-key loop iterates old_state_dict, so a new-only head reached no comparison at all. It is not specific to the key this PR adds either: lm_head.weight and model.language_model.lm_head.weight were already excused and already unchecked in that direction.

Each new-only tied head is now compared against the embedding it is tied to, resolved as lm_head -> embed_tokens on the same prefix before falling back to TIED_EMBED_KEYS, at the same atol=1e-4 / rtol=1e-3 as the existing tied fallback. When the old side carries no embedding to compare against it stays silent rather than inventing a failure.

Covered by test_a_new_only_tied_head_is_checked_against_its_embedding (all four tied names) plus test_a_new_only_head_with_no_embedding_to_check_against_is_not_a_failure. The four new cases fail on the parent commit and pass here, and the 17-case matrix (agrees / disagrees / both sides, each direction) is green.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
